airlock/file_browser_api.py

```diff
@@ -95,24 +110,218 @@ def html_classes(self):
+    root_node = PathItem(
+        container=release_request,
+        relpath=ROOT_PATH,
+        type="request",
+        parent=None,
+        **is_selected(ROOT_PATH, selected_path),
+    )
+
+    for name, group in release_request.filegroups.items():
+        group_path = UrlPath(name)
+        group_node = PathItem(
+            container=release_request,
+            relpath=group_path,
+            type="filegroup",
+            parent=root_node,
+            display_text=f"{name} ({len(group.files)} files)",
+            **is_selected(group_path, selected_path),
+        )
+
+        group_node.children = get_filegroup_tree(
+            release_request,
+            selected_path,
+            group,
+            group_node.relpath,
+            parent=group_node,
+        )
+
+        root_node.children.append(group_node)
+
+    return root_node
+
+
+def get_filegroup_tree(container, selected_path, group_data, group_path, parent):
+    """Get the tree for a filegroup's files.
+
+    This is more than just a walk the disk. The FileGroup.files is flat list of
+    relative paths. So we need group those by common prefex and descend down
+    the tree.
+    """
+
+    def build_filegroup_tree(file_parts, path, parent):
+        """Walk a flat list of paths and create a diretcories tree for them."""
+
+        # group multiple paths into groups by first part of path
+        grouped = defaultdict(list)
+        for child, *descendants in file_parts:
+            if descendants:
+                grouped[child].append(descendants)
+            else:
+                grouped[child] = []
+
+        tree = []
+
+        # now we have them grouped by first path element, we can great a node
+        # in the tree for them
+        for child, descendants in grouped.items():
+            child_path = path / child
+
+            node = PathItem(
+                container=container,
+                relpath=child_path,
+                parent=parent,
+                # actual path on disk, striping the group part
+                filepath=child_path.relative_to(child_path.parts[0]),
+                **is_selected(child_path, selected_path),
+            )
+
+            abspath = node._absolute_path()
+            assert abspath.exists()
+
+            if descendants:
+                assert abspath.is_dir()
+                node.type = "dir"
+                # recurse down the tree
+                node.children = build_filegroup_tree(
+                    descendants, child_path, parent=node
+                )
+            else:
+                assert abspath.is_file()
+                node.type = "file"
+
+            tree.append(node)
+
+        # sort directories first then files
+        tree.sort(key=children_sort_key)
+        return tree
+
+    file_parts = [f.relpath.parts for f in group_data.files]
+    return build_filegroup_tree(file_parts, group_path, parent)
+
+
+def is_selected(path, selected_path):
```

Comment:
This was confusing because a PathItem also has an `is_selected` method, and because I was expecting it to return a boolean (so unpacking it was odd).

Reply:
yeah, it is not clean.

But there are basically two attributes to set based on the selected_path, and I didn't want to have to repeat the logic each time we create a PathItem, and this is what I came up with.

The alternative as a post-instiation `set_selected(item, selected_path)` function I guess? I thought this was neater.

But I also wrote it when PathItem was still a frozen dataclass. I had to make it mutable in the end, so that the children could have a reference to their parent, before the parent's children had been set.

I'll have another look and see if there's a better way to do this.

Reply:
Have moved to a `set_selected(node, selected_path)` pattern.

Reply:
tbh I think just naming it differently would have unconfused me, but this works :+1: